File: android/app/src/main/kotlin/com/example/poc_platform_view/OffshoreFxWrapper.kt
package com.example.poc_platform_view

import android.content.Context
import android.os.Looper
import android.os.Handler
import android.view.View
import com.saulo.dev.offshorefx.OffshoreFxModel
import com.saulo.dev.offshorefx.OffshoreFxView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class OffshoreFxWrapper(
    context: Context, id: Int,
    messenger: BinaryMessenger,
    creationParams: Map<String, Any>?
) : PlatformView, OffshoreFxView.OffshoreFxCallback {

    private val methodChannel = MethodChannel(messenger, "offshore_fx_channel")

    private val offshoreFxView = OffshoreFxView(context)

    private val uiHandler = Handler(Looper.getMainLooper())

    override fun getView(): View = offshoreFxView

    init {
        offshoreFxView.setOffshoreFxCallback(this)

        OffshoreFxModel.fromMap(creationParams)?.let {
            data -> offshoreFxView.initWithData(data)
        }
    }

    override fun dispose() {

        offshoreFxView.clearFocus()
        offshoreFxView.cleanup()

        val parent = offshoreFxView.parent
        if (parent is android.view.ViewGroup) {
            parent.removeView(offshoreFxView)
        }
    }

    // --- Implementação da Interface OffshoreFxCallback ---
    override fun onSuccess(data: Map<String, Any>) {
        uiHandler.post {
            methodChannel.invokeMethod("onSuccess", data)
        }

    }

    override fun onError(errorMessage: String) {
        uiHandler.post {
            methodChannel.invokeMethod("onError", mapOf("message" to errorMessage))
        }

    }

    override fun onPaused() {
        uiHandler.post {
            methodChannel.invokeMethod("onPaused", null)
        }

    }
}

class OffshoreFxFactory(private val messenger: BinaryMessenger) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(
        context: Context,
        viewId: Int,
        args: Any?
    ): PlatformView {
        val params = args as? Map<String, Any>?
        return OffshoreFxWrapper(
            context,
            viewId,
            messenger,
            params,
        )
    }
}

File: ios/Runner/OffshoreFxWrapper.swift
//
//  OffshoreFxWrapper.swift
//  Runner
//
//  Created by Saulo Nascimento on 17/01/26.
//

import Flutter
import UIKit
import OffshoreFx

class OffshoreFxWrapper: NSObject, FlutterPlatformView, OffshoreFxCallback {
    private var _view: OffshoreFxView
    private let _channel: FlutterMethodChannel
    
    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger
    ) {
        _view = OffshoreFxView(frame: frame)
        _channel = FlutterMethodChannel(name: "offshore_fx_channel", binaryMessenger: messenger)
        
        super.init()
        
        _view.callback = self
        
        
        if let params = args as? [String: Any] {
            if let model = OffshoreFxModel.fromMap(params) {
                _view.initWithData(model)
            }
        }
    }
    
    func view() -> UIView { return _view }
    
    // MARK: - OffshoreFxCallback
    func onSuccess(data: [String: Any]) {
        _channel.invokeMethod("onSuccess", arguments: data)
    }
    
    func onError(errorMessage: String) {
        _channel.invokeMethod("onError", arguments: ["message": errorMessage])
    }
    
    func onPaused() {
        _channel.invokeMethod("onPaused", arguments: nil)
    }
}

class OffshoreFxFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger
    init(messenger: FlutterBinaryMessenger) { self.messenger = messenger }
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        return OffshoreFxWrapper(frame: frame, viewIdentifier: viewId, arguments: args, binaryMessenger: messenger)
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}
